File: backend/app/schemas/book.py
from datetime import date, datetime
from typing import Literal
from pydantic import BaseModel, Field, ConfigDict


class BookImageBase(BaseModel):
    url: str = Field(..., min_length=5, max_length=500)
    caption: str | None = None


class BookImageCreate(BookImageBase):
    pass


class BookImageOut(BookImageBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    created_at: datetime


class ReadingLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    book_id: int
    pages_read: int
    current_page_after: int
    logged_at: datetime


class BookBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    author: str = Field(..., min_length=1, max_length=255)
    total_pages: int = Field(..., gt=0)
    start_date: date | None = None
    notes: str | None = None
    description: str | None = None
    cover_url: str | None = None


class BookCreate(BookBase):
    current_page: int = Field(default=0, ge=0)
    finish_date: date | None = None


class BookUpdate(BaseModel):
    title: str | None = None
    author: str | None = None
    total_pages: int | None = Field(default=None, gt=0)
    current_page: int | None = Field(default=None, ge=0)
    start_date: date | None = None
    finish_date: date | None = None
    notes: str | None = None
    description: str | None = None
    cover_url: str | None = None


class BookOut(BookBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    current_page: int
    finish_date: date | None
    created_at: datetime
    updated_at: datetime

    progress_percent: float = 0.0
    daily_average: float = 0.0
    estimated_finish_date: date | None = None
    status: str = "in_progress"

    images: list[BookImageOut] = []
    logs: list[ReadingLogOut] = []


class DashboardStats(BaseModel):
    total_books: int
    in_progress: int
    completed: int
    average_daily_pages: float
    overall_progress: float


class PeriodStats(BaseModel):
    start: date
    end: date
    days: int
    books_completed: int
    books_completed_list: list[BookOut]
    pages_logged: int          # net (positivos + negativos)
    pages_logged_gross: int    # apenas positivos
    pages_corrections: int     # |soma dos negativos|
    pages_total_from_completed: int
    average_per_day: float     # baseado em net / days


class TimeseriesPoint(BaseModel):
    bucket: str
    pages: int           # net
    pages_gross: int     # somente positivos
    sessions: int


class TimeseriesStats(BaseModel):
    start: date
    end: date
    granularity: Literal["year", "month", "week"]
    total_pages: int
    points: list[TimeseriesPoint]

File: frontend/src/types/book.ts
export interface BookImage {
  id: number;
  url: string;
  caption: string | null;
  created_at: string;
}

export interface ReadingLog {
  id: number;
  book_id: number;
  pages_read: number;
  current_page_after: number;
  logged_at: string;
}

export interface Book {
  id: number;
  title: string;
  author: string;
  total_pages: number;
  current_page: number;
  start_date: string | null;
  finish_date: string | null;
  cover_url: string | null;
  description: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  progress_percent: number;
  daily_average: number;
  estimated_finish_date: string | null;
  status: "not_started" | "in_progress" | "completed";
  images: BookImage[];
  logs: ReadingLog[];
}

export interface BookCreate {
  title: string;
  author: string;
  total_pages: number;
  current_page?: number;
  start_date?: string | null;
  finish_date?: string | null;
  notes?: string | null;
  cover_url?: string | null;
}

export interface DashboardStats {
  total_books: number;
  in_progress: number;
  completed: number;
  average_daily_pages: number;
  overall_progress: number;
}

export interface PeriodStats {
  start: string;
  end: string;
  days: number;
  books_completed: number;
  books_completed_list: Book[];
  pages_logged: number;          // net
  pages_logged_gross: number;    // só positivos
  pages_corrections: number;     // |soma dos negativos|
  pages_total_from_completed: number;
  average_per_day: number;
}

export type Granularity = "year" | "month" | "week";

export interface TimeseriesPoint {
  bucket: string;
  pages: number;        // net
  pages_gross: number;  // só positivos
  sessions: number;
}

export interface TimeseriesStats {
  start: string;
  end: string;
  granularity: Granularity;
  total_pages: number;
  points: TimeseriesPoint[];
}
